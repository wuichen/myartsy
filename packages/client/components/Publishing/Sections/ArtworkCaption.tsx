import { color as Color, Sans, Serif } from "@artsy/palette"
import { ErrorBoundary } from "Components/ErrorBoundary"
import { pMedia } from "Components/Helpers"
import { ArticleLayout, SectionLayout } from "Components/Publishing/Typings"
import { Truncator } from "Components/Truncator"
import _ from "lodash"
import React from "react"
import styled from "styled-components"

interface ArtworkCaptionProps extends React.HTMLProps<HTMLDivElement> {
  artwork: any
  color?: string
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
  linked?: boolean
  isFullscreenCaption?: boolean
}

export class ArtworkCaption extends React.Component<ArtworkCaptionProps> {
  static defaultProps = {
    color: Color("black30"),
  }

  joinParts(children, key, delimiter = ", ") {
    const compacted = _.compact(children)

    if (compacted.length) {
      const reduced = compacted.reduce((prev, curr, i) => {
        return [
          prev,
          <span key={`joinParts-${key}-${i}`}>{delimiter}</span>,
          curr,
        ]
      })
      return reduced
    } else {
      return []
    }
  }

  joinArtistNames(names, delimiter = ", ") {
    if (names.length === 0) {
      return []
    }

    return names.slice(1).reduce(
      (prev, curr, i) => {
        return prev.concat([
          <span key={`joinArtistNames-${i}`}>{delimiter}</span>,
          curr,
        ])
      },
      [names[0]]
    )
  }

  renderArtists() {
    const {
      artwork: { artist, artists },
    } = this.props

    // Multiple artists
    if (artists && artists.length > 1) {
      const names = artists.map((a, i) => {
        const artistName = this.renderArtistName(a, `renderArtists-${i}`)
        return artistName
      })

      const joinedNames = this.joinArtistNames(names)
      return joinedNames

      // Single artist
    } else if (artist) {
      const artistName = this.renderArtistName(artist, "renderArtists-single")
      return artistName
    }
  }

  renderArtistName(artist, key: string) {
    const { linked } = this.props
    const { name, slug } = artist
    const createTextLink = linked && slug

    if (createTextLink) {
      const href = `/artist/${slug}`

      return (
        <ArtistName key={`renderArtistName-${key}`}>
          <a href={href}>{name}</a>
        </ArtistName>
      )
    } else {
      return (
        <span key={`renderArtistName-${key}`} className="name">
          {name}
        </span>
      )
    }
  }

  renderTitleDate() {
    const children = [this.renderTitle(), this.renderDate()]
    const titleDate = this.joinParts(children, "renderTitleDate")
    return titleDate
  }

  renderTitle() {
    const {
      artwork: { slug, title },
      linked,
    } = this.props

    if (title) {
      if (linked) {
        const href = `/artwork/${slug}`

        return (
          <span key="renderTitle" className="title">
            <a href={href}>{title}</a>
          </span>
        )
      } else {
        return (
          <span key="renderTitle" className="title">
            {title}
          </span>
        )
      }
    }
  }

  renderDate() {
    const {
      artwork: { date },
    } = this.props

    if (date && date.length) {
      return (
        <span key="renderDate" className="date">
          {date}
        </span>
      )
    }
  }

  renderPartner() {
    const {
      artwork: {
        partner: { name, slug },
      },
      linked,
    } = this.props

    if (name) {
      const createTextLink = Boolean(linked && slug)

      if (createTextLink) {
        return (
          <a key="renderPartner" href={`/${slug}`}>
            {name}
          </a>
        )
      } else {
        return name
      }
    }
  }

  renderCredit() {
    const {
      artwork: { credit },
    } = this.props

    if (credit && credit.length) {
      return (
        <span key="renderCredit" className="credit">
          {credit}
        </span>
      )
    }
  }

  renderPartnerCredit = () => {
    const children = [this.renderPartner(), this.renderCredit()]

    const joined = this.joinParts(children, "renderPartnerCredit", ". ")
    return joined
  }

  renderFullscreenCaption = () => {
    return (
      <StyledFullscreenCaption size={["3", "4"]} weight="medium">
        <Line>
          <ArtistNames>{this.renderArtists()}</ArtistNames>
        </Line>
        <div>
          <Line>{this.renderTitleDate()}</Line>
          <Line>{this.renderPartnerCredit()}</Line>
        </div>
      </StyledFullscreenCaption>
    )
  }

  renderClassicCaption = () => {
    return (
      <StyledClassicCaption
        color={Color("black60")}
        size="2"
        className="display-artwork__caption"
      >
        <Truncator>
          <ArtistNames>{this.renderArtists()}</ArtistNames>
          {this.renderTitleDate()}
          {". "}
          {this.renderPartner()}
        </Truncator>
      </StyledClassicCaption>
    )
  }

  renderEditorialCaption = () => {
    const { color, layout, sectionLayout } = this.props

    return (
      <StyledArtworkCaption
        size="3"
        color={color}
        layout={layout}
        sectionLayout={sectionLayout}
      >
        <ArtistNames>{this.renderArtists()}</ArtistNames>
        <div>
          <Truncator>{this.renderTitleDate()}</Truncator>
          <Truncator>{this.renderPartnerCredit()}</Truncator>
        </div>
      </StyledArtworkCaption>
    )
  }

  render() {
    const { layout, isFullscreenCaption } = this.props

    return (
      <ErrorBoundary>
        <div>
          {isFullscreenCaption
            ? this.renderFullscreenCaption()
            : layout === "classic"
            ? this.renderClassicCaption()
            : this.renderEditorialCaption()}
        </div>
      </ErrorBoundary>
    )
  }
}

const ArtistNames = styled.span`
  margin-right: 30px;
`

const ArtistName = styled.span`
  white-space: nowrap;
`

export const StyledArtworkCaption = styled(Sans)<{
  color?: string
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
}>`
  padding: ${props => (props.sectionLayout === "fillwidth" ? "0 10px;" : "0;")};
  margin-top: 10px;
  display: flex;

  a {
    color: ${props => props.color};
    text-decoration: none;
  }

  .title,
  .title a {
    font-style: italic;
  }

  ${pMedia.xs`
    padding: 0 10px;
  `};
`

const StyledClassicCaption = styled(Serif)<{
  className?: string
  color?: string
}>`
  margin-top: 10px;
  display: block;

  a {
    color: ${props => props.color};
    text-decoration: none;
  }

  ${ArtistNames} {
    margin-right: 0;
    font-weight: bold;

    &::after {
      content: ", ";
    }
  }

  .title {
    font-style: italic;
  }
`

const StyledFullscreenCaption = styled(Sans)`
  display: flex;

  a {
    color: black;
    text-decoration: none;
  }

  .title,
  .title a {
    font-style: italic;
  }

  ${pMedia.sm`
    flex-direction: column;
  `};
`

const Line = styled.div`
  ${pMedia.sm`
    &.artist-name {
      margin-bottom: 5px;
    }
  `};
`
