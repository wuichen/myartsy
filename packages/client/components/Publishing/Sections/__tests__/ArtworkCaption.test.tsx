import {
  ArtworkMissingInfo,
  ArtworkMultipleArtists,
  ArtworkRegular,
} from "Components/Publishing/Fixtures/Components"
import { mount } from "enzyme"
import "jest-styled-components"
import _ from "lodash"
import React from "react"
import renderer from "react-test-renderer"
import { ArtworkCaption } from "../ArtworkCaption"

describe("ArtworkCaption", () => {
  const getWrapper = (props: any = {}) => {
    return mount(<ArtworkCaption artwork={props.artwork || ArtworkRegular} />)
  }

  describe("snapshot", () => {
    it("renders a fullscreen caption properly", () => {
      const caption = renderer
        .create(<ArtworkCaption artwork={ArtworkRegular} isFullscreenCaption />)
        .toJSON()
      expect(caption).toMatchSnapshot()
    })

    it("renders a classic caption properly", () => {
      const caption = renderer
        .create(<ArtworkCaption artwork={ArtworkRegular} layout="classic" />)
        .toJSON()
      expect(caption).toMatchSnapshot()
    })

    it("renders a standard caption properly", () => {
      const caption = renderer
        .create(<ArtworkCaption artwork={ArtworkRegular} layout="standard" />)
        .toJSON()
      expect(caption).toMatchSnapshot()
    })
  })

  describe("#joinParts", () => {
    it("joins zero items", () => {
      const component = getWrapper() as any
      const joined = component.instance().joinParts([])
      expect(joined.toString()).toEqual("")
    })

    it("joins one item", () => {
      const component = getWrapper() as any
      const joined = component.instance().joinParts(["Title"])
      expect(joined.toString()).toEqual("Title")
    })

    it("joins two items", () => {
      const component = getWrapper() as any
      const joined = component.instance().joinParts(["Title", "Date"])

      expect(joined[0]).toEqual("Title")
      expect(joined[2]).toEqual("Date")
    })

    it("joins three items into a nested array", () => {
      const component = getWrapper() as any
      const joined = component
        .instance()
        .joinParts(["Title", "Date", "Partner"])
      expect(joined.length).toEqual(3)
      expect(joined[0][0]).toEqual("Title")
      expect(joined[0][2]).toEqual("Date")
      expect(joined[2]).toEqual("Partner")
    })
  })

  describe("render methods", () => {
    it("can render with missing information", () => {
      const component = getWrapper({
        artwork: ArtworkMissingInfo,
      })
      expect(component.text().length).toBe(0)
    })

    it("renders a single artist", () => {
      const component = getWrapper()
      expect(component.html()).toMatch("Fernando Botero")
    })

    it("renders artists", () => {
      const component = getWrapper({
        artwork: _.extend({}, ArtworkMultipleArtists, {
          artists: [{ name: "Andy Warhol" }, { name: "Botero" }],
        }),
      })
      expect(component.text()).toMatch("Andy Warhol, Botero")
    })

    it("renders title + date", () => {
      const component = getWrapper({
        artwork: ArtworkRegular,
      })
      expect(component.text()).toMatch("Nude on the Beach, 2000")
    })

    it("renders partner + credit", () => {
      const component = getWrapper()
      expect(component.text()).toMatch("Gary Nader. Courtesy of Gary Nader")
    })
  })
})
