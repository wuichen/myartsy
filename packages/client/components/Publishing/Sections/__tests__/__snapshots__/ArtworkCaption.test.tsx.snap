// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtworkCaption snapshot renders a classic caption properly 1`] = `
.c1 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 14px;
  line-height: 18px;
  color: #666;
}

.c0 {
  margin-top: 10px;
  display: block;
}

.c0 a {
  color: #666;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 .c2 {
  margin-right: 0;
  font-weight: bold;
}

.c0 .c2::after {
  content: ", ";
}

.c0 .title {
  font-style: italic;
}

<div>
  <div
    className="display-artwork__caption c0 c1"
    color="#666"
    fontFamily="'Adobe Garamond W08', 'adobe-garamond-pro', 'AGaramondPro-Regular', 'Times New Roman', Times, serif"
    fontSize="14px"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<span><span class=\\"c2 ciyMlo\\"><span class=\\"name\\">Fernando Botero</span></span><span class=\\"title\\">Nude on the Beach</span><span>, </span><span class=\\"date\\">2000</span>. Gary Nader</span>",
        }
      }
    />
  </div>
</div>
`;

exports[`ArtworkCaption snapshot renders a fullscreen caption properly 1`] = `
.c1 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
}

.c3 {
  margin-right: 30px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 a {
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 .title,
.c0 .title a {
  font-style: italic;
}

@media screen and (min-width:40em) {
  .c1 {
    font-size: 16px;
  }
}

@media screen and (min-width:40em) {
  .c1 {
    line-height: 26px;
  }
}

@media (max-width:720px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:720px) {
  .c2.artist-name {
    margin-bottom: 5px;
  }
}

<div>
  <div
    className="c0 c1"
    fontFamily={
      Object {
        "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
        "fontWeight": 500,
      }
    }
    fontSize={
      Array [
        "14px",
        "16px",
      ]
    }
  >
    <div
      className="c2"
    >
      <span
        className="c3"
      >
        <span
          className="name"
        >
          Fernando Botero
        </span>
      </span>
    </div>
    <div>
      <div
        className="c2"
      >
        <span
          className="title"
        >
          Nude on the Beach
        </span>
        <span>
          , 
        </span>
        <span
          className="date"
        >
          2000
        </span>
      </div>
      <div
        className="c2"
      >
        Gary Nader
        <span>
          . 
        </span>
        <span
          className="credit"
        >
          Courtesy of Gary Nader
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ArtworkCaption snapshot renders a standard caption properly 1`] = `
.c1 {
  font-family: Unica77LLWebRegular,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #C2C2C2;
}

.c3 {
  margin-right: 30px;
}

.c0 {
  padding: 0;
  margin-top: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 a {
  color: #C2C2C2;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 .title,
.c0 .title a {
  font-style: italic;
}

.c4 .c2 {
  margin-right: 0;
  font-weight: bold;
}

.c4 .c2::after {
  content: ", ";
}

@media (max-width:600px) {
  .c0 {
    padding: 0 10px;
  }
}

<div>
  <div
    className="c0 c1"
    color="#C2C2C2"
    fontFamily="Unica77LLWebRegular, 'Helvetica Neue', Helvetica, Arial, sans-serif"
    fontSize="14px"
  >
    <span
      className="c2 c3"
    >
      <span
        className="name"
      >
        Fernando Botero
      </span>
    </span>
    <div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span><span class=\\"title\\">Nude on the Beach</span><span>, </span><span class=\\"date\\">2000</span></span>",
          }
        }
      />
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span>Gary Nader<span>. </span><span class=\\"credit\\">Courtesy of Gary Nader</span></span>",
          }
        }
      />
    </div>
  </div>
</div>
`;
